docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+```
+#!/bin/bash
+(...)
+echo "#devfile-status# {'buildStatus':'Compiling application'}
+mvn compile
+echo "#devfile-status# {'buildStatus':'Running unit tests'}
+mvn test
+```
+
+This 'detailed build status' markup text is *entirely optional*: if this markup is not present, the odo tool can still determine build succeeded/failed and build running/not-running using the other `odo push -o json` JSON events. 
+
+### App status notification via `odo component status -o json` and `odo log --follow`
+
+In general, within the execution context that odo operates, there are a few ways for us to determine the application status:
+1) Can the application be pinged at its exposed URL? 
+2) What state is the container in? (running/container creating/restarting/etc -- different statuses between local and Kube but same general idea)
+3) In the application log, specific hardcoded text strings can be searched for (for example, OpenLiberty outputs defined status codes to its log to indicate that an app started.)
+
+Ideally, we would like for odo to provide consuming tools with all 3 sets of data. Thus, as proposed:
+- 1 and 2 are handled by a new `odo component status -o json` command, described here.
+- 3 is handled by the existing unmodified `odo log --follow` command.
+
+The new proposed `odo component status -o json` command will:
+- Be a *long-running* command that will continue outputing status until it is aborted by the parent process.
+- Every X seconds, ping the URLs/routes of the application as they existed when the command was first executed. Output the result as a JSON string.
+- Every X seconds (or using a Kubernetes watch, where appropriate), check the container status for the application, based on the application data that was present when the command was first issued. Output the result as a JSON string.
+
+**Note**: This command will NOT, by design, respond to route/application changes that occur during or after it is first invoked. It is up to consuming tools to ensure that the `odo component status` command is stopped/restarted as needed. 
+  - For example, if the user tells the IDE to delete their application with the IDE UI, the IDE will call `odo delete (component)`; at the same time, the IDE should also abort any existing `odo component status` commands that are running (as these are no longer guaranteed to return a valid status now that the application itself no longer exists). `odo component status` will not automatically abort when the application is deleted (because it has no reliable way to detect this in all cases).
+  - Another example: if the IDE adds a new URL via `odo url create [...]`, any existing `odo component status` commands that are running should be aborted, as these commands would still only be checking the URLs that existed when the command was first invoked (eg there is intentionally no cross-process notification mechanism for created/updated/deleted URLs implemented as part of this command.)
```

Comment:
This answers my previous long comment about a hypothetical scenario about change in URL/app to a good extent. However, I think the last point in that comment referring to `odo deploy` still stands. Keeping the onus on IDE, or any other tooling that's consuming odo stream, to stop the `odo component status` process sounds like a deal-breaker (apologies, couldn't find a milder word).

Instead, if `odo component status -o json` command can be designed to be smart enough to handle these changes in URL/app, it might be a pain for us initially but, the users would find it to be the most useful command once they decide to deploy their application using odo. Something like "odo Day 2". :wink:

Reply:
Re: the 'deal-breaker' terminology, no worries 😄. Hopefully I can convince you otherwise... I have added [a new section to the proposal where I examine some of the other solutions I considered](https://github.com/jgwest/odo/blob/doc-2550-proposal/docs/proposals/event-notification.md#other-solutions-considered).

See that new section for details, but in short, the option I propose here -- requiring the IDE to restart the odo status process-- is admittedly not a perfect solution, but is less complex (and better meets our other requirements) than the alternatives.

Certainly we certainly COULD go with any of the other more sophisticated options: there are no technical impediments to doing so, and speaking personally it would be interesting/fun to implement them 😃 . 

BUT, I suspect this is a case where the perfect is the enemy of the good: it would take an order of magnitude greater time/effort to implement a more sophisticated solution, and in the mean time (with IDEs champing at the bit to add support for odo dev files) the existing requirement to provide app/build status doesn't go away: we would still need a short-term stop-gap measure which would probably look a lot like this proposal anyways.

Finally, re: Day 2, this option definitely does not in any way tie our hands in implementing a more complex solution in the future if/when they become necessary for 'day two' odo usage. This proposal is strictly scoped to current requirements: the `odo component status` command is ONLY designed to be invoked by IDEs (or other similar tools) that want to provide detailed app/build status to a user through a UI. It is not designed for users to invoke themselves (for example, from deployment scripts). 

But in a future day 2 proposal we could definitely look at using the building blocks discussed here to build command behaviour like that, ideally as part of a larger deployment story.

Reply:
> See that new section for details, but in short, the option I propose here -- requiring the IDE to restart the odo status process -- is admittedly not a perfect solution, but is less complex (and better meets our other requirements) than the alternatives.

Agreed.

> BUT, I suspect this is a case where the perfect is the enemy of the good:

+1. I have the negative trait of perfectionism. Doesn't mean I don't write buggy code - I do that all the time. :wink: But when I imagined the day 2 scenario and the "onus on IDE" part, I think it was my perfectionist point of view at play.

> Finally, re: Day 2, this option definitely does not in any way tie our hands in implementing a more complex solution in the future if/when they become necessary for 'day two' odo usage.

Most important thing is that it leaves the room open for further development around the ideas I had while still having a functionality that can be useful to the users (here, IDE plugins) fairly quickly.

Among the proposed solutions, the one about `fsnotify` makes more sense than client-server architecture. Of course they both come with their own pros & cons but, that's for another day when we plan to handle the complex stuff.

Thanks for this discussion and sharing more ideas! :+1:

Reply:
> BUT, I suspect this is a case where the perfect is the enemy of the good: it would take an order of magnitude greater time/effort to implement a more sophisticated solution, and in the mean time (with IDEs champing at the bit to add support for odo dev files) the existing requirement to provide app/build status doesn't go away: we would still need a short-term stop-gap measure which would probably look a lot like this proposal anyways.

+1

I think that what is currently being proposed where tools need to restart `odo component status` after the component was modified is ok. If we are talking about IDE plugins, they will safely know when the `odo component status` restart will be required.